### PR TITLE
os/bluestore: SMR updates/cleanups

### PR DESCRIPTION
**Pull request by liewegas**
lots of stuff here, enough to make ceph_test_objectstore pass (mostly) against a simulated zoned device with a separate conventional db device.  notably,

- allocator/freelist setup/teardown is cleaned up
- zone refs / backpointers are reimplemented
- write and clone paths are fully implemented
- fsck
- simple cleaner

### CODE REVIEW

**Comment by liewegas** [comment]
This looks ok in rados suite; these failures are unrelated to these changes.
https://pulpito.ceph.com/sage-2021-09-16_18:04:19-rados-wip-sage-testing-2021-09-16-1020-distro-basic-smithi/

File: src/test/objectstore/store_test.cc [hunk 1]
```diff
@@ -276,6 +278,15 @@ TEST_P(StoreTest, TrivialRemount) {
   ASSERT_EQ(0, r);
 }
 
+TEST_P(StoreTest, TrivialRemountFsck) {
+  int r = store->umount();
+  ASSERT_EQ(0, r);
+  r = store->fsck(false);
+  ASSERT_EQ(0, r);
```
Comment (neha-ojha) [inline src/test/objectstore/store_test.cc:285]:
```
2021-09-17T04:52:51.001 INFO:teuthology.orchestra.run.smithi174.stdout:[ RUN      ] ObjectStore/StoreTest.TrivialRemountFsck/1
2021-09-17T04:52:51.763 INFO:teuthology.orchestra.run.smithi174.stdout:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-7765-g05d4cac9/rpm/el8/BUILD/ceph-17.0.0-7765-g05d4cac9/src/test/objectstore/store_test.cc:285: Failure
2021-09-17T04:52:51.764 INFO:teuthology.orchestra.run.smithi174.stdout:Expected equality of these values:
2021-09-17T04:52:51.764 INFO:teuthology.orchestra.run.smithi174.stdout:  0
2021-09-17T04:52:51.764 INFO:teuthology.orchestra.run.smithi174.stdout:  r
2021-09-17T04:52:51.765 INFO:teuthology.orchestra.run.smithi174.stdout:    Which is: -95
```

https://pulpito.ceph.com/sage-2021-09-16_18:04:19-rados-wip-sage-testing-2021-09-16-1020-distro-basic-smithi/6393227/

**Comment by neha-ojha** [comment]
> This looks ok in rados suite; these failures are unrelated to these changes.
> https://pulpito.ceph.com/sage-2021-09-16_18:04:19-rados-wip-sage-testing-2021-09-16-1020-distro-basic-smithi/

@liewegas I am seeing many occurrences of https://tracker.ceph.com/issues/52138, which has now been fixed, in your run. Could you include the fix in your next run?

File: src/test/objectstore/store_test.cc (hunk 1 above)
Comment (liewegas) [reply src/test/objectstore/store_test.cc:287]: fixed

**Comment by liewegas** [comment]
> > This looks ok in rados suite; these failures are unrelated to these changes.
> > https://pulpito.ceph.com/sage-2021-09-16_18:04:19-rados-wip-sage-testing-2021-09-16-1020-distro-basic-smithi/
> 
> @liewegas I am seeing many occurrences of https://tracker.ceph.com/issues/52138, which has now been fixed, in your run. Could you include the fix in your next run?

The last run included it (as well as most of the ceph-volume fixes): https://pulpito.ceph.com/sage-2021-09-20_04:54:37-rados-wip-sage-testing-2021-09-19-2154-distro-basic-smithi/

looks ok to me?

**Comment by neha-ojha** [comment]
> > > This looks ok in rados suite; these failures are unrelated to these changes.
> > > https://pulpito.ceph.com/sage-2021-09-16_18:04:19-rados-wip-sage-testing-2021-09-16-1020-distro-basic-smithi/
> > 
> > 
> > @liewegas I am seeing many occurrences of https://tracker.ceph.com/issues/52138, which has now been fixed, in your run. Could you include the fix in your next run?
> 
> The last run included it (as well as most of the ceph-volume fixes): https://pulpito.ceph.com/sage-2021-09-20_04:54:37-rados-wip-sage-testing-2021-09-19-2154-distro-basic-smithi/
> 
> looks ok to me?

Yeah, I don't think any of those are related. I also did a rerun of the failed jobs, https://pulpito.ceph.com/nojha-2021-09-20_21:24:58-rados-wip-sage-testing-2021-09-19-2154-distro-basic-smithi/ - the upgrade test failure should be fixed by https://github.com/ceph/ceph/pull/43211 and opened https://tracker.ceph.com/issues/52676 for the standalone test failure.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -555,6 +563,37 @@ static int get_key_pool_stat(const string& key, uint64_t* pool_id)
...
+
+static int get_key_zone_offset_object(
+  const string& key,
+  uint32_t *zone,
+  uint64_t *offset,
+  ghobject_t *oid)
+{
+  const char *p = key.c_str();
+  if (key.length() < sizeof(uint64_t) + sizeof(uint32_t))
+    return -1;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:587]: ```if (key.length() < sizeof(uint64_t) + sizeof(uint32_t) + ENCODED_KEY_PREFIX_LEN + 1)```
_get_key_object() needs p to contain at least ENCODED_KEY_PREFIX_LEN+1 characters.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -5475,71 +5534,149 @@ int BlueStore::_write_out_fm_meta(uint64_t target_size)
...
+    vector<uint64_t> wp = bdev->get_zones();
+    vector<zone_state_t> zones = f->get_zone_states(db);
+    ceph_assert(wp.size() == zones.size());
+
+    // reconcile zone state
+    auto num_zones = bdev->get_size() / zone_size;
+    map<uint32_t, uint64_t> zone_adjustments;
+    for (unsigned i = first_sequential_zone; i < num_zones; ++i) {
+      uint64_t p = wp[i] == (i + 1) * zone_size ? zone_size : wp[i] % zone_size;
+      if (zones[i].write_pointer > p) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5612]: More checks and simpler calculation:
```
ceph_assert(wp[i] >= i * zone_size);
ceph_assert(wp[i] <= (i + 1) * zone_size);
uint64_t p = wp[i] - i * zone_size;
```

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -5475,71 +5534,149 @@ int BlueStore::_write_out_fm_meta(uint64_t target_size)
...
+
+    if (!read_only && !zone_adjustments.empty()) {
+      dout(1) << __func__ << " adjusting freelist based on device write pointers" << dendl;
+      KeyValueDB::Transaction t = db->get_transaction();
+      for (auto& i : zone_adjustments) {
+	// note that the offset is imprecise, but only need to select the zone
+	f->allocate(i.first * zone_size, i.second, t);
+      }
+      db->submit_transaction_sync(t);
+    }
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5656]: We invoke `_init_alloc()` from `_open_db_and_around()`, in the part where db is always opened in read-only.
The value _read_only_ is misleading at this point - when *false* it just means that in future we will reopen db in r/w mode.

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -6298,7 +6447,7 @@ void BlueStore::_dump_alloc_on_failure()
     cct->_conf->bluestore_bluefs_alloc_failure_dump_interval;
   if (dump_interval > 0 &&
     next_dump_on_bluefs_alloc_failure <= ceph_clock_now()) {
-    shared_alloc.a->dump();
+    alloc->dump();
     next_dump_on_bluefs_alloc_failure = ceph_clock_now();
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:6451]: `_dump_alloc_on_failure()` is dead code now, but I think it should refer `shared_alloc.a`, not BlueStore Block (private) data.

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -9072,34 +9378,84 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
...
+	uint64_t pos = used_blocks.find_first();
+
+	// skip conventional zones
+	while (pos != decltype(used_blocks)::npos &&
+	       (pos * min_alloc_size) / zone_size < first_sequential_zone) {
+	  dout(40) << " (conventional) zone 0x" << std::hex
+		   << (pos * min_alloc_size) / zone_size
+		   << " dead 0x" << (pos * min_alloc_size) << "~" << min_alloc_size
+		   << std::dec << dendl;
+	  pos = used_blocks.find_next(pos);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:9399]: I see ZonedAllocator is allocating only from sequential zones.
Should presence of used_blocks in conventional zones be considered an error?

File: src/os/bluestore/BlueStore.cc (hunk 4 above)
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:5656]: the code as it stands is correct, though, right?  the first (read-only) pass won't adjust the freelist's write pointer, but the second (!read-only) pass will?

File: src/os/bluestore/BlueStore.cc (hunk 6 above)
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:9399]: Perhaps, but I don't want to invest too much effort here since I think this is likely to change.  (1) There are partial bluefs changes that allow it to consume sequential zones. (2) We probably want bluefs to spill over into sequential zones as needed for sst files, but use the conventional region for its log files.  (3) We could conceivably have bluestore put some objects in the conventional zone if the hints indicate a high likelihood of overwrites.  If we go down any of these paths, then we probably need a single Allocator that will pull from one region or the other based on some flag.

File: src/os/bluestore/BlueStore.cc (hunk 4 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:5658]: To be sure I added test.
https://github.com/ceph/ceph/pull/43257
It fails on DB when write pointer needs to be fixed.

Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:5658]: Ah, I see it now!  Thanks for that test

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -14662,6 +14697,29 @@ void BlueStore::_wctx_finish(
...
+	  zones_with_releases.erase(e.offset / zone_size);
+	}
+      }
+    }
+    for (auto zone : zones_with_releases) {
+      auto p = o->onode.zone_offset_refs.find(zone);
+      if (p != o->onode.zone_offset_refs.end()) {
+	dout(20) << __func__ << " rm ref zone 0x" << std::hex << zone
+		 << " offset 0x" << p->second << std::dec << dendl;
+	txc->note_release_zone_offset(o, zone, p->second);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:15178]: this function call erases from onode's zone_offset_refs[zone] which can be done more efficiently here using the p iterator.

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:15178]: additionally as both zones_with_releases and zone_offset_refs are sorted by zone one can use more efficient algorithm to locate the zone in _refs: 
for each zone in zons_with_releases
  scan in _refs from the previous position while current < zone
  if current == zone  
    note_release_zone
    current++
  previous = current

File: src/os/bluestore/ZonedAllocator.cc [hunk 8]
```diff
@@ -151,7 +151,7 @@ void ZonedAllocator::dump(std::function<void(uint64_t offset,
 }
 
 void ZonedAllocator::init_from_zone_pointers(
-  std::vector<zone_state_t> &&_zone_states,
+  std::vector<zone_state_t> _zone_states,
```
Comment (ifed01) [inline src/os/bluestore/ZonedAllocator.cc]: any rationales for that change?

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -564,20 +564,33 @@ static int get_key_pool_stat(const string& key, uint64_t* pool_id)
 }
 
 #ifdef HAVE_LIBZBD
-static void get_zone_offset_key(uint32_t zone, uint64_t offset, std::string *key)
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc]: The idea was to make the references sorted, so that we would process objects to clean in (approximate) order they appear on disk.  In retrospect, though, I think we should consider just reading the whole zone into memory in one go, or possibly use a granular freelist so we read just the live extents.  Even if we read a bunch of dead data, on a spinning disk that is likely to be way more efficient than seeking at least once for each live object.  Or perhaps scan the set of live objects to build the live extent list, read them, then scan again to rewrite and update the object metadata.  None of those options need the offset.

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -13041,11 +13061,136 @@ void BlueStore::_zoned_cleaner_thread()
...
+    _clean_some(oid, zone);
+  }
+
+  if (a->get_live_bytes(zone) > 0) {
+    derr << "zone 0x" << std::hex << zone << " still has 0x" << a->get_live_bytes(zone)
+	 << " live bytes" << std::dec << dendl;
+    // should we do something else here to avoid a live-lock in the event of a problem?
+    return;
+  }
+
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc:13099]: need to drain transactions

File: src/os/bluestore/BlueStore.cc [hunk 11]
```diff
@@ -5610,6 +5629,12 @@ int BlueStore::_init_alloc(std::map<uint64_t, uint64_t> *zone_adjustments)
       }
     }
 
+    // start with conventional zone "free" (bluefs may adjust this when it starts up)
+    auto reserved = _get_ondisk_reserved();
+    shared_alloc.a->init_add_free(reserved,
+				  bdev->get_conventional_region_size() - reserved);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:5635]: Shouldn't this use the same 
p2align(bdev->get_conventional_region_size(), min_alloc_size) - reserved
pattern for the length as mkfs does?

**Comment by ifed01** [comment]
Just a side note about using conventional space for BlueFS. 
Perhaps that's not the best idea at all - BlueFS tend to perform append-only operations so wouldn't it be better to use sequential partition for it only?

File: src/os/bluestore/ZonedAllocator.cc [hunk 12]
```diff
@@ -148,44 +150,26 @@ void ZonedAllocator::dump(std::function<void(uint64_t offset,
...
-
-  for ( ; length; length -= zone_size) {
-    increment_write_pointer(++zone_num, zone_size);
-    ldout(cct, 40) << " set zone 0x" << std::hex
-		   << zone_num << " write pointer to 0x" << zone_size << std::dec << dendl;
+  ldout(cct, 10) << dendl;
+  cleaner_lock = _cleaner_lock;
+  cleaner_cond = _cleaner_cond;
+  zone_states = std::move(_zone_states);
+  num_free = 0;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc:159]: renaming num_free -> num_sequential_free and adjusted calculations

File: src/os/bluestore/ZonedAllocator.cc (hunk 8 above)
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc]: this used to get passed directly from ``ZonedFreelistManager::get_zone_states()``.  the caller is done with it at this point so I can std::move, though, so doing that.

File: src/os/bluestore/BlueStore.cc [hunk 13]
```diff
@@ -5561,8 +5561,27 @@ int BlueStore::_create_alloc()
     return -EINVAL;
   }
 
-  // BlueFS will share the same allocator
-  shared_alloc.set(alloc);
+#ifdef HAVE_LIBZBD
+  if (freelist_type == "zoned") {
+    Allocator *a = Allocator::create(
+      cct, cct->_conf->bluestore_allocator,
+      bdev->get_conventional_region_size(),
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:5568]: adding an assert.  for now we still need somewhere to put the superblock.  we can deal with ZNS drives later..

File: src/os/bluestore/BlueStore.cc [hunk 14]
```diff
@@ -4538,8 +4538,8 @@ static void discard_cb(void *priv, void *priv2)
 void BlueStore::handle_discard(interval_set<uint64_t>& to_release)
 {
   dout(10) << __func__ << dendl;
-  ceph_assert(shared_alloc.a);
-  shared_alloc.a->release(to_release);
+  ceph_assert(alloc);
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:4541]: i think we actually want to skip discard always in the smr (or zns) case(s)...

File: src/os/bluestore/BlueStore.cc [hunk 15]
```diff
@@ -9167,77 +9167,160 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
...
+	uint64_t zone_dead = 0;
+	for (uint32_t zone = first_sequential_zone;
+	     zone < num_zones;
+	     ++zone, zone_dead = 0) {
+	  while (pos != decltype(used_blocks)::npos &&
+		 (pos * min_alloc_size) / zone_size == zone) {
+	    dout(40) << " zone 0x" << std::hex << zone
+		     << " dead 0x" << (pos * min_alloc_size) << "~" << min_alloc_size
+		     << std::dec << dendl;
+	    zone_dead += min_alloc_size;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:9422]: the loop above marks everything from wp to zone end as used, so after the flip the unwritten portion won't appear at all.

File: src/os/bluestore/ZonedAllocator.cc [hunk 16]
```diff
@@ -110,8 +102,6 @@ int64_t ZonedAllocator::allocate(
 		 << " and zone offset 0x" << (offset % zone_size)
 		 << std::dec << dendl;
 
-  find_zones_to_clean();
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc:105]: yeah. in practice it wakes up anyway for whatever reason.  i'm not sure what policy we want to pursue here.. but for now I'll just make it a timed sleep.

File: src/os/bluestore/BlueStore.cc [hunk 17]
```diff
@@ -8671,6 +8670,30 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
     dout(1) << __func__ << " debug abort" << dendl;
     goto out_scan;
   }
+
+#ifdef HAVE_LIBZBD
+  if (bdev->is_smr()) {
+    auto a = dynamic_cast<ZonedAllocator*>(shared_alloc.a);
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc]: yeah, i'm not sure how to make this nicer. the bottom line is that the zoned freelist *can't* actually enumerate meaningfully.. it can only make up something with the same dead and unwritten byte count.

**Comment by liewegas** [comment]
> Just a side note about using conventional space for BlueFS. Perhaps that's not the best idea at all - BlueFS tend to perform append-only operations so wouldn't it be better to use sequential partition for it only?

Mostly yes.  For sst files that is certainly true.  The challenge is the bluefs log, which overwrites the trailing block multiple times.  One way to fix this would be to make the .log files have a special format in which all writes on disk write complete blocks, but actually contain a length wrapper around them, so that the logical offset doesn't directly map to the ondisk offset.  This would complicate the read path etc.  In practice, for now, it's simplest to just defer those bluefs changes until later.  We have the conventional region so we may as well use it (at least for the .log files).

File: src/os/bluestore/ZonedAllocator.cc [hunk 18]
```diff
@@ -167,25 +167,45 @@ void ZonedAllocator::init_from_zone_pointers(
...
 {
   int32_t best = -1;
-  int64_t best_score = 0;
+  float best_score = 0.0;
   for (size_t i = first_seq_zone_num; i < num_zones; ++i) {
-    int64_t score = zone_states[i].num_dead_bytes;
-    // discount by remaining space so we will tend to clean full zones
-    score -= (zone_size - zone_states[i].write_pointer) / 2;
-    if (score > 0 && (best < 0 || score > best_score)) {
+    int64_t live_bytes = zone_states[i].write_pointer - zone_states[i].num_dead_bytes;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc:175]: leaving this as-is since get_live_bytes() may need to be changed to include locking

File: src/os/bluestore/ZonedAllocator.cc [hunk 19]
```diff
@@ -167,25 +167,45 @@ void ZonedAllocator::init_from_zone_pointers(
 		 << dendl;
 }
 
-int64_t ZonedAllocator::pick_zone_to_clean(void)
+int64_t ZonedAllocator::pick_zone_to_clean(float min_score, uint64_t min_saved)
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc:170]: added locking

File: src/os/bluestore/BlueStore.cc [hunk 20]
```diff
@@ -7549,6 +7752,21 @@ BlueStore::OnodeRef BlueStore::fsck_check_objects_shallow(
...
 
+#ifdef HAVE_LIBZBD
+    if (bdev->is_smr() && depth != FSCK_SHALLOW) {
+      for (auto& e : blob.get_extents()) {
+	if (e.is_valid()) {
+	  uint32_t zone = e.offset / zone_size;
+	  uint64_t offset = e.offset % zone_size;
+	  auto p = zone_first_offsets.find(zone);
+	  if (p == zone_first_offsets.end() || p->second > offset) {
+	    zone_first_offsets[zone] = offset;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:7763]: nit: suboptimal implementation - can reuse previous lookup result, see https://stackoverflow.com/questions/97050/stdmap-insert-or-stdmap-find
Mind leaving a FIXME?

File: src/os/bluestore/BlueStore.cc [hunk 21]
```diff
@@ -7667,6 +7885,33 @@ BlueStore::OnodeRef BlueStore::fsck_check_objects_shallow(
...
+		 << "; this can happen due to clone_range"
+		 << dendl;
+	  } else {
+	    dout(20) << " good zone ref 0x" << std::hex << zone << " offset 0x" << p->second
+		     << " <= first offset 0x" << first_offset
+		     << std::dec << dendl;
+	  }
+	  (*ctx.zone_refs)[zone].erase(p);
+	} else {
+	  derr << "fsck error: " << oid << " references zone 0x" << std::hex << zone
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:7907]: this looks easily repairable  at this point by inserting proper record under PREFIX_ZONED_CL_INFO prefix. Mark with TODO?

File: src/os/bluestore/BlueStore.cc [hunk 22]
```diff
@@ -8560,6 +8875,19 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
...
   }
 
+#ifdef HAVE_LIBZBD
+  if (bdev->is_smr() && depth != FSCK_SHALLOW) {
+    dout(1) << __func__ << " checking for leaked zone refs" << dendl;
+    for (uint32_t zone = 0; zone < zone_refs.size(); ++zone) {
+      for (auto& [oid, offset] : zone_refs[zone]) {
+	derr << "fsck error: stray zone ref 0x" << std::hex << zone
+	     << " offset 0x" << offset << " -> " << std::dec << oid << dendl;
+	++errors;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:8885]: Mark with TODO - looks repairable

File: src/os/bluestore/BlueStore.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:9418]: Right

File: src/os/bluestore/ZonedAllocator.cc [hunk 23]
```diff
@@ -124,143 +133,109 @@ void ZonedAllocator::release(const interval_set<uint64_t>& release_set) {
...
+    float score =
+      (float)zone_states[i].num_dead_bytes /
+      (float)(zone_states[i].get_num_live_bytes() + 1);
+    if (score > 0) {
+      ldout(cct, 20) << " zone 0x" << std::hex << i
+		     << " dead 0x" << zone_states[i].num_dead_bytes
+		     << " score " << score
+		     << dendl;
+    }
+    if (zone_states[i].num_dead_bytes < min_saved) {
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/ZonedAllocator.cc:189]: i have it there so you can can still see the debug line if we skip the zone

File: src/os/bluestore/BlueStore.cc [hunk 24]
```diff
@@ -5231,10 +5274,15 @@ void BlueStore::_set_alloc_sizes(void)
 {
   max_alloc_size = cct->_conf->bluestore_max_alloc_size;
 
+#ifdef HAVE_LIBZBD
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc:5277]: this is called from handle_conf_change

**Comment by liewegas** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc [hunk 25]
```diff
@@ -17965,7 +18569,7 @@ int BlueStore::add_existing_bluefs_allocation(Allocator* allocator, read_alloc_s
     }
     for (auto itr = bluefs_extents.begin(); itr != bluefs_extents.end(); extent_count++, itr++) {
       //dout(5) << "BlueFS[" << extent_count << "] <" << itr.get_start() << "," << itr.get_len() << ">" << dendl;
-      allocator->init_rm_free(itr.get_start(), itr.get_len());
+      shared_alloc.a->init_rm_free(itr.get_start(), itr.get_len());
```
Comment (benhanokh) [inline src/os/bluestore/BlueStore.cc:18572]: Please revert the change as it breaks the function.
The function modifies a temp @allocator and should not operate on the global shared-allocator

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:18572]: @benhanokh 
> The function modifies a temp @Allocator and should not operate on the global shared-allocator

Please fire a relevant ticket for this bug - commenting on a merged PR isn't a proper way of dealing with the issue.
